src/crimson/os/seastore/omap_manager/btree/omap_btree_node_impl.cc

```diff
@@ -0,0 +1,643 @@
+
+std::ostream &operator<<(std::ostream &out, const omap_inner_key_t &rhs)
+{
+  return out << "omap_inner_key (" << rhs.key_off<< " - " << rhs.key_len
+             << " - " << rhs.laddr << ")";
+}
+
+std::ostream &operator<<(std::ostream &out, const omap_leaf_key_t &rhs)
+{
+  return out << "omap_leaf_key_t (" << rhs.key_off<< " - " << rhs.key_len
+             << " "<< rhs.val_off<<" - " << rhs.val_len << ")";
+}
+
+std::ostream &OMapInnerNode::print_detail_l(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << get_meta().depth;
+}
+
+OMapInnerNode::handle_split_ret
+OMapInnerNode::handle_split(omap_context_t oc, internal_iterator_t iter,
+                               mutation_result_t mresult)
+{
+  logger().debug("{}: {}","OMapInnerNode",  __func__);
+  auto [left, right, pivot] = *(mresult.split_tuple);
+  if (!extent_will_overflow(pivot.size() + 1, std::nullopt)) {
+    if (!is_pending()) {
+      auto mut = oc.tm.get_mutable_extent(oc.t, this)->cast<OMapInnerNode>();
+      auto mut_iter = mut->iter_idx(iter.get_index());
+      return mut->handle_split(oc, mut_iter, mresult);
+    }
+    journal_inner_update(iter, left->get_laddr(), maybe_get_delta_buffer());
+    journal_inner_insert(iter + 1, right->get_laddr(), pivot,
+                         maybe_get_delta_buffer());
+    return insert_ret(
+           insert_ertr::ready_future_marker{},
+           mutation_result_t(mutation_status_t::SUCCESS, std::nullopt, std::nullopt));
+  } else {
+    return make_split_children(oc)
+      .safe_then([this, oc, iter, mresult = std::move(mresult)] (auto tuple) {
+       //parent node left and right
+       auto [p_left, p_right, p_pivot] = tuple;
+       //child node left and right
+       auto [c_left, c_right, c_pivot] = *(mresult.split_tuple);
+       if (c_pivot < p_pivot) {  //left
+         auto mut_iter = p_left->iter_idx(iter.get_index());
+         p_left->journal_inner_update(mut_iter, c_left->get_laddr(),
+                                      p_left->maybe_get_delta_buffer());
+         p_left->journal_inner_insert(mut_iter + 1, c_right->get_laddr(), c_pivot,
+                                      p_left->maybe_get_delta_buffer());
+       } else {  //right
+         auto mut_iter = p_right->iter_idx(iter.get_index() - p_left->get_node_size());
+         p_right->journal_inner_update(mut_iter, c_left->get_laddr(),
+                                       p_right->maybe_get_delta_buffer());
+         p_right->journal_inner_insert(mut_iter + 1, c_right->get_laddr(), c_pivot,
+                                         p_right->maybe_get_delta_buffer());
+       }
+       return oc.tm.dec_ref(oc.t, get_laddr())
+         .safe_then([tuple = std::move(tuple)] (auto ret) {
+          return insert_ret(
+                 insert_ertr::ready_future_marker{},
+                 mutation_result_t(mutation_status_t::SPLITTED, tuple, std::nullopt));
+       });
```

Comment:
The point of my comment before is that this logic and the logic you factored out into make_replace are almost identical.  Can you generalize make_replace (and rename it to just 'replace') to handle this case as well?

Reply:
handle_splitted, operations are update and insert. make_replace operations are update and replace.  they are different operations. so do you mean I check mresult.status, if SPLITTED go a branch and then go another branch?

Reply:
Split a into a' and b:
- remove a,
- insert a'
- insert b

Full Merge a, b into c:
- remove a
- remove b
- insert c

Balance Merge a, b into a', b':
- remove a
- remove b
- insert a'
- insert b'

I don't see enough of a difference between update, replace, and insert to justify duplicating the logic.  I think you can simplify it to a range of iterators to remove and a sequence of key/addr pairs to insert.  You walk the removal range to get the size to remove, sum the set to be inserted to get the size to insert, and then check size to insert - size to remove to check whether a split is necessary.

Reply:
Something like

```
/**
 * inner_replace
 *
 * Replaces entries in [begin, end) with the contents of replacements.  May result in a split or merge outcome
 * encoded in the returned mutation_result_t
 */
ertr::future<mutation_result_t>
inner_replace(iterator begin, iterator end, std::vector<std::pair<std::string, laddr_t>> replacements);
```

Reply:
You can recover the update optimization (in all three cases, begin->key() remains invariant) by adding a precondition to inner_replace that begin != end, begin->get_key() == replacements.front().first.

Reply:
I have a problem here, for split, I need compare key(parent pivot and child pivot) , since there is a special case parent node exactly split at  these two iterators, and child left and child right always should put to parent left.  but for make balance, at this special case, I need check iterator index == left size, can't use pivot compare,  any suggestion here?

Reply:
That's just an artifact of how make_split_children works, right?  How about passing begin, end, replacements into make_split_children and constructing the children directly using the new keys.  make_split_children should be able to compute the final size and split up the keys (including the new ones) appropriately.

Reply:
I seems have a solution. let me try.

Reply:
I do update and remove in old node, then when old node do make_split_children, the content already updated only need do  insert in new left node or right node, so that I can use pivot to comapre.  PR already updated, please check make_split_insert function and the its caller.

Reply:
It would be helpful to me if you could do the rebase, @liu-chunmei .

Reply:
I will rebase after no much need to be modified.

Reply:
@mattbenjamin What's blocked on rebasing this PR?